File: integration-tests/src/test/java/pl/grzeslowski/capybara/test/HigherOrderFunctionsTest.java
package pl.grzeslowski.capybara.test;

import org.junit.jupiter.api.Test;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

class HigherOrderFunctionsTest {
    @Test
    void mapWithFunctionReference() {
        assertThat(HigherOrderFunctions.map(List.of(1, 2, 3)))
                .isEqualTo(List.of(1, 4, 9));
    }

    @Test
    void invokesHigherOrderFunction() {
        assertThat(HigherOrderFunctions.hof(x -> "n=" + x))
                .isEqualTo("n=5");
    }

    @Test
    void invokesHigherOrderFunctionWithLambdaCapture() {
        assertThat(HigherOrderFunctions.invokeHof("Hello"))
                .isEqualTo("Hello 5");
    }

    @Test
    void invokesHigherOrderFunctionWithMultiArgumentLambda() {
        assertThat(HigherOrderFunctions.runLambda())
                .isEqualTo(-45);
    }

    @Test
    void invokesHigherOrderFunctionWithFunctionReference() {
        assertThat(HigherOrderFunctions.runLambda2())
                .isEqualTo(-45);
    }

    @Test
    void invokesHigherOrderFunctionWithUnusedSingleArgument() {
        assertThat(HigherOrderFunctions.invokeHofUnused("Hello"))
                .isEqualTo("Hello");
    }

    @Test
    void invokesHigherOrderFunctionWithUnusedMultiArguments() {
        assertThat(HigherOrderFunctions.runLambdaUnusedMiddle())
                .isEqualTo(12);
    }
}

File: capy/src/e2e-cfun/js/dev/capylang/test/HigherOrderFunctions.test.js
const test = require('node:test');
const { assert, generatedModule } = require('./_support');

test('HigherOrderFunctions', () => {
    const higher = generatedModule('dev/capylang/test/HigherOrderFunctions.js');
    assert.deepEqual(higher.map([1, 2, 3]).asList(), [1, 4, 9]);
    assert.equal(higher.hof(x => `n=${x}`), 'n=5');
    assert.equal(higher.invokeHof('Hello'), 'Hello 5');
    assert.equal(higher.runLambda(), -45);
    assert.equal(higher.runLambda2(), -45);
    assert.equal(higher.invokeHofUnused('Hello'), 'Hello');
    assert.equal(higher.runLambdaUnusedMiddle(), 12);
    assert.equal(higher.runLambdaNoArgs(), 'Hello, Wrold');
});
